doc/interpreter/mkoctfile.1
.\" --------------------------------------------------------------------
.\"
.\" Copyright (C) 2000-2026 The Octave Project Developers
.\"
.\" See the file COPYRIGHT.md in the top-level directory of this
.\" or <https://octave.org/copyright/>.
.\"
.\" This file is part of Octave.
.\"
.\" Octave is free software: you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" Octave is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with Octave; see the file COPYING.  If not, see
.\" <https://www.gnu.org/licenses/>.
.\"
.\" --------------------------------------------------------------------
.\"
.de Vb \" (V)erbatim (b)egin.  Use fixed width font and no justification
.ft CW
.nf
..
.de Ve \" (V)erbatim (e)nd.  Return to regular font and justification
.ft R
.fi
..
.\" --------------------------------------------------------------------
.TH mkoctfile 1 "April 2026" "GNU Octave"
.SH NAME
mkoctfile \- Compile dynamic\-load modules for GNU Octave
.SH SYNOPSIS
\fBmkoctfile\fP [\fIoption\fP]\|.\|.\|.\& FILE\|.\|.\|.\&
.SH DESCRIPTION
\fBmkoctfile\fP is used to compile C, C++, or Fortran source code in to a
dynamically loadable \fB.oct\fP file for
.MR octave 1 .
.SH OPTIONS
.TP
\fB\-h\fP, \fB\-\-help\fP
Print help message.
.TP
.B \-I\fIdir\fP
Add include directory \fIdir\fP to compile command.
.TP
.B \-idirafter \fIdir\fP
Add include directory to second include search path for compilation
after '\fB\-I\fP'.
.TP
.B \-D\fIdef\fP
Add definition \fIdef\fP to compile command.
.TP
.B \-l\fIlib\fP
Add library \fIlib\fP to link command.
.TP
.B \-L\fIdir\fP
Add library directory \fIdir\fP to link command.
.TP
.B \-M\fR,\fB \-\-depend
Generate dependency files (\fB.d\fP) for C and C++ source files.
.TP
.B \-pthread
Add \fB\-pthread\fP to link command.
.TP
.B \-R\fIdir\fP
Pass \fB\-R\fP\fIdir\fP to link command.
.TP
.B \-Wl,\fIoption\fP
Pass option to the linker such as \fB\-Wl,\-rpath=...\fP
.TP
.B \-W...
Pass flags to the compiler such as \fB\-Wa,\fP\fIoption\fP.
.TP
.B \-c\fR,\fB \-\-compile
Compile, but do not link.
.TP
.B \-o \fIfile\fP\fR,\fB \-\-output \fIfile\fP
Output filename.  Default extension is "\fB.oct\fP" (or "\fB.mex\fP" if
\fB\-\-mex\fP is specified) unless linking a stand-alone executable.
.TP
.B \-g
Enable debugging option for all compilers.
.TP
.B \-p \fIVAR\fP\fR,\fB \-\-print \fIVAR\fP
Print configuration variable \fIVAR\fP.  There are three categories of
variables:
.IP
Octave configuration variables that users may override with environment
variables.  These are used in commands that \%\fBmkoctfile\fP executes.

.RS
.Vb
    ALL_CFLAGS                  INCLUDEDIR
    ALL_CXXFLAGS                LAPACK_LIBS
    ALL_FFLAGS                  LDFLAGS
    ALL_LDFLAGS                 LD_STATIC_FLAG
    BLAS_LIBS                   LFLAGS
    CC                          LIBDIR
    CFLAGS                      LIBOCTAVE
    CPICFLAG                    LIBOCTINTERP
    CPPFLAGS                    OCTAVE_LINK_OPTS
    CXX                         OCTINCLUDEDIR
    CXXFLAGS                    OCTAVE_LIBS
    CXXLD                       OCTAVE_LINK_DEPS
    CXXPICFLAG                  OCTLIBDIR
    DL_LDFLAGS                  OCT_LINK_DEPS
    F77                         OCT_LINK_OPTS
    F77_INTEGER8_FLAG           RDYNAMIC_FLAG
    FFLAGS                      SPECIAL_MATH_LIB
    FPICFLAG                    XTRA_CFLAGS
    INCFLAGS                    XTRA_CXXFLAGS
.Ve
.RE
.IP
Octave configuration variables as above, but currently unused by
\%\fBmkoctfile\fP.

.RS
.Vb
    AR
    DEPEND_EXTRA_SED_PATTERN
    DEPEND_FLAGS
    FFTW3F_LDFLAGS
    FFTW3F_LIBS
    FFTW3_LDFLAGS
    FFTW3_LIBS
    FFTW_LIBS
    FLIBS
    LIBS
    RANLIB
    READLINE_LIBS
.Ve
.RE
.IP
Octave configuration variables that are provided for informational purposes
only.  Except for \%\fIOCTAVE_HOME\fP and \%\fIOCTAVE_EXEC_HOME\fP, users may
not override these variables.
.IP
If \%\fIOCTAVE_HOME\fP or \%\fIOCTAVE_EXEC_HOME\fP are set in the environment,
then other variables are adjusted accordingly with \%\fIOCTAVE_HOME\fP or
\%\fIOCTAVE_EXEC_HOME\fP substituted for the original value of the directory
specified by the \%\-\-\fIprefix\fP or \%\-\-\fIexec\-prefix\fP options that
were used when Octave was configured.

.RS
.Vb
    API_VERSION                 LOCALARCHLIBDIR
    ARCHLIBDIR                  LOCALFCNFILEDIR
    BINDIR                      LOCALOCTFILEDIR
    CANONICAL_HOST_TYPE         LOCALSTARTUPFILEDIR
    DATADIR                     LOCALVERARCHLIBDIR
    DATAROOTDIR                 LOCALVERFCNFILEDIR
    DEFAULT_PAGER               LOCALVEROCTFILEDIR
    EXEEXT                      MAN1DIR
    FCNFILEDIR                  MAN1EXT
    IMAGEDIR                    MANDIR
    INCLUDEDIR                  OCTAVE_EXEC_HOME
    INFODIR                     OCTAVE_HOME
    INFOFILE                    OCTDATADIR
    LIBDIR                      OCTDOCDIR
    LIBEXECDIR                  OCTFILEDIR
    LOCALAPIARCHLIBDIR          OCTFONTSDIR
    LOCALAPIFCNFILEDIR          OCTINCLUDEDIR
    LOCALAPIOCTFILEDIR          OCTLIBDIR
    LOCALAPIPKGDIR              STARTUPFILEDIR
                                VERSION
.Ve
.RE
.TP
.B \-\-link\-stand\-alone
Link a stand-alone executable file.
.TP
.B \-\-mex
Create a MEX file.  Set the default output extension to "\fB.mex\fP".  Link to
\%liboctmex instead of \%liboctinterp and \%liboctave.
.TP
.B \-s\fR,\fB \-\-strip
Strip output file.
.TP
.B \-n\fR,\fB \-\-dry-run\fR,\fB \-\-just-print\fR
Print commands, but do not execute them.
.TP
.B \-v\fR,\fB \-\-verbose
Echo commands as they are executed.
.TP
.B FILE
Compile or link FILE.  Recognized file types are:

.RS
.Vb
   .c    C source
   .cc   C++ source
   .cp   C++ source
   .cpp  C++ source
   .CPP  C++ source
   .cxx  C++ source
   .c++  C++ source
   .C    C++ source
   .f    Fortran source (fixed form)
   .F    Fortran source (fixed form)
   .f90  Fortran source (free form)
   .F90  Fortran source (free form)
   .o    object file
   .a    library file
.Ve
.RE
.SH AUTHOR
John W. Eaton and many others.  The list of contributors to the Octave project
may be shown from within \%\fBoctave\fP by executing
"\fB\%doc \%Acknowledgements\fP".  The list is also available online at
.UR https://\:\%docs.octave.org/\:\%latest/\:\%Acknowledgements\:\%.html
.UE .
.SH "SEE ALSO"
.MR octave 1 .


NEWS
Summary of important user-visible changes for version 7 (yyyy-mm-dd):
----------------------------------------------------------------------

### General improvements

- Many functions in Octave can be called in a command form---no
parentheses for invocation and no return argument assignment---or in a
functional form---parentheses and '=' for assignment of return values.

    **Command Form Example**

    `mkdir new_directory`

    **Function Form Example**

    `status = mkdir ("new_directory")`

    Octave now handles errors that occur in a consistent manner.  If
    called in command form and there is a failure, an error is thrown
    and a message printed.  If called in functional form, no error or
    message is printed and the failure is communicated to the programmer
    via the output status variable.

    The following list of functions have been modified.

    * `copyfile`
    * `fcntl`
    * `fileattrib`
    * `kill`
    * `link`
    * `mkfifo`
    * `movefile`
    * `rename`
    * `rmdir`
    * `symlink`
    * `unlink`

- Calling a user-defined function with too many inputs or outputs is now
an error.  The interpreter makes this check automatically.  If a
function uses varargin then the check is skipped for function inputs,
and if a function uses varargout then the check is skipped for function
outputs.  Input validation for functions typically begins with checking
that the number of inputs and outputs match expectations.  Existing code
can be simplified by removing these checks which are now done by the
interpreter.  Typically, code blocks like the following can simply be
deleted.

        ## Checking number of inputs
        if (nargin > 2)
          print_usage ();
        endif

        ## Checking number of outputs
        if (nargout > 1)
          print_usage ();
        endif

- Binary and hexadecimal constants like `0b101` and `0xDEADBEEF` now
create integers (unsigned by default) with sizes determined from the
number of digits present.  For example, `0xff` creates a `uint8` value
and `0xDEADBEEF` creates a `uint64` value.  You may also use a suffix of
the form `u8`, `u16`, `u32`, `u64`, `s8`, `s16`, `s32`, or `s64` to
explicitly specify the data type to use (`u` or `s` to indicate unsigned
or signed and the number to indicate the integer size).

    Binary constants are limited to 64 binary digits and hexadecimal
constants are limited to 16 hexadecimal digits with no automatic
rounding or conversion to floating point values.  Note that this may
cause problems in existing code.  For example, an expression like
`[0x1; 0x100; 0x10000]` will be uint8 (because of the rules of
concatenating integers of different sizes) with the larger values
truncated (because of the saturation semantics of integer values).  To
avoid these kinds of problems either: 1) declare the first integer to be
of the desired size such as `[0x1u32; 0x100; 0x10000]`, or 2) pad
constants in array expressions with leading zeros so that they use the
same number of digits for each value such as
`[0x00_00_01; 0x00_01_00; 0x01_00_00]`.

- The increment and decrement operators `++` and `--` must "hug" their
corresponding variables.  In previous versions of Octave, whitespaces
between these operators and the variable they affect were allowed.  That
is no longer the case.

- The `mldivide` function (i.e., the `\` operator) now uses an LU
decomposition to solve nearly singular full square matrices.  This is
Matlab-compatible and yields results which more nearly minimize `norm
(A*x - b)`.  Previously, Octave computed a minimum-norm solution.

- The `factor` function has been overhauled for speed.  For large
inputs > 1e14, it can be up to 10,000 times faster.

- The `isprime` function uses a new primality testing algorithm
that is up to 50,000 times faster for inputs > 1e14.

- The `betainc` function now calculates an exact output for the
important special cases where a or b are 1.

- The `whos` function now displays an additional attribute 's' when
the variable is a sparse type.

- As part of GSoC 2020, Abdallah K. Elshamy implemented the
`jsondecode` and `jsonencode` functions to read and write JSON data.

- As part of GSoC 2021, Abdallah K. Elshamy implemented the
`jupyter_notebook` classdef class.  This class supports running and
filling Jupyter Notebooks using the Octave language kernel from Octave
itself.  Making the evaluation of long-running Jupyter Notebooks on a
computing server without permanent browser connection possible.

- By default, the history file is now located at $DATA/octave/history,
where $DATA is a platform dependent location for (roaming) user data
files (e.g., ${XDG_DATA_HOME} or, if that is not set, ~/.local/share on
Unix-like operating systems or %APPDATA% on Windows).

- For Octave on Windows OS, the minimum required version of the Windows
API is now 6.1 (Windows 7 or newer).

- The non-re-entrant version of the QHull library "libqhull" was
deprecated upstream.  Octave now (optionally) requires the re-entrant
version of that library "libqhull_r" instead.

- Octave's libraries are now built using symbol visibility by default.
That means that less symbols are exported from these libraries.
Configure with `--disable-lib-visibility-flags` to export all symbols
(as in previous versions).

- Octave's build system no longer appends "++" to the end of the
"magick++" library name (set with the `--with-magick=` configure flag).
The real name of the "magick++" library (including any potentially
trailing "++") needs to be set in that option now.

- The `pkg update` command now accepts options that are passed to `pkg
install` for each updated package.  Specifying @option{-local} or
@option{-global} will restrict update checks to local or global
installed packages, respectively.

### Graphical User Interface

- The graphical user interface is now available in Hungarian and
Turkish.

- In debug mode, symbol values are now shown in tooltips when hovering
variables in the editor panel.


scripts/miscellaneous/mkoctfile.m
########################################################################
##
## Copyright (C) 2006-2026 The Octave Project Developers
##
## See the file COPYRIGHT.md in the top-level directory of this
## distribution or <https://octave.org/copyright/>.
##
## This file is part of Octave.
##
## Octave is free software: you can redistribute it and/or modify it
## under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
##
## Octave is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with Octave; see the file COPYING.  If not, see
## <https://www.gnu.org/licenses/>.
##
########################################################################

## -*- texinfo -*-
## @deftypefn  {} {} mkoctfile [-options] file @dots{}
## @deftypefnx {} {[@var{output}, @var{status}] =} mkoctfile (@dots{})
##
## The @code{mkoctfile} function compiles source code written in C, C++, or
## Fortran.  Depending on the options used with @code{mkoctfile}, the
## compiled code can be called within Octave or can be used as a stand-alone
## application.
##
## @code{mkoctfile} can be called from the shell prompt or from the Octave
## prompt.  Calling it from the Octave prompt simply delegates the call to
## the shell prompt.  Any output is stored in the @var{output} variable and
## the exit status in the @var{status} variable.  If called with no outputs
## and the compilation fails then Octave will emit an error.  If the programmer
## requests @var{output} or @var{status}, however, Octave will merely issue
## a warning and it is the programmer's responsibility to verify the command
## was successful.
##
## @code{mkoctfile} accepts the following options, all of which are optional
## except for the filename of the code you wish to compile:
##
## @table @samp
## @item -I DIR
## Add the include directory DIR to compile commands.
##
## @item -D DEF
## Add the definition DEF to the compiler call.
##
## @item -l LIB
## Add the library LIB to the link command.
##
## @item -L DIR
## Add the library directory DIR to the link command.
##
## @item  -M
## @itemx --depend
## Generate dependency files (.d) for C and C++ source files.
##
## @item -R DIR
## Add the run-time path to the link command.
##
## @item @nospell{-Wl,@dots{}}
## Pass options to the linker like @nospell{"-Wl,-rpath=@dots{}"}.
## The quotes are needed since commas are interpreted as command
## separators.
##
## @item -W@dots{}
## Pass options to the assembler like @nospell{"-Wa,OPTION"}.
##
## @item -c
## Compile but do not link.
##
## @item -g
## Enable debugging options for compilers.
##
## @item  -o FILE
## @itemx --output FILE
## Output filename.  Default extension is @file{.oct} (or @file{.mex} if
## @samp{--mex} is specified) unless linking a stand-alone executable.
##
## @item  -p VAR
## @itemx --print VAR
## Print configuration variable VAR@.  There are three categories of
## variables:
##
## Octave configuration variables that users may override with environment
## variables.  These are used in commands that @code{mkoctfile} executes.
##
## @example
##    ALL_CFLAGS                  INCLUDEDIR
##    ALL_CXXFLAGS                LAPACK_LIBS
##    ALL_FFLAGS                  LDFLAGS
##    ALL_LDFLAGS                 LD_STATIC_FLAG
##    BLAS_LIBS                   LIBDIR
##    CC                          LIBOCTAVE
##    CFLAGS                      LIBOCTINTERP
##    CPICFLAG                    LIBOCTMEX
##    CPPFLAGS                    OCTAVE_LINK_OPTS
##    CXX                         OCTINCLUDEDIR
##    CXXFLAGS                    OCTAVE_LIBS
##    CXXLD                       OCTAVE_LINK_DEPS
##    CXXPICFLAG                  OCTLIBDIR
##    DL_LDFLAGS                  OCT_LINK_DEPS
##    F77                         OCT_LINK_OPTS
##    F77_INTEGER8_FLAG           RDYNAMIC_FLAG
##    FFLAGS                      SPECIAL_MATH_LIB
##    FPICFLAG                    XTRA_CFLAGS
##    INCFLAGS                    XTRA_CXXFLAGS
## @end example
##
## Octave configuration variables as above, but currently unused by
## @code{mkoctfile}.
##
## @example
## @group
##    AR
##    DEPEND_EXTRA_SED_PATTERN
##    DEPEND_FLAGS
##    FFTW3F_LDFLAGS
##    FFTW3F_LIBS
##    FFTW3_LDFLAGS
##    FFTW3_LIBS
##    FFTW_LIBS
##    FLIBS
##    LIBS
##    RANLIB
##    READLINE_LIBS
## @end group
## @end example
##
## Octave configuration variables that are provided for informational
## purposes only.  Except for @samp{OCTAVE_HOME} and @samp{OCTAVE_EXEC_HOME},
## users may not override these variables.
##
## If @w{@env{OCTAVE_HOME}}@ or @w{@env{OCTAVE_EXEC_HOME}}@ are set in the
## environment, then other variables are adjusted accordingly with
## @w{@env{OCTAVE_HOME}}@ or @w{@env{OCTAVE_EXEC_HOME}}@ substituted for the
## original value of the directory specified by the @option{--prefix} or
## @option{--exec-prefix} options that were used when Octave was configured.
##
## @example
## @group
##    API_VERSION                 LOCALARCHLIBDIR
##    ARCHLIBDIR                  LOCALFCNFILEDIR
##    BINDIR                      LOCALOCTFILEDIR
##    CANONICAL_HOST_TYPE         LOCALSTARTUPFILEDIR
##    DATADIR                     LOCALVERARCHLIBDIR
##    DATAROOTDIR                 LOCALVERFCNFILEDIR
##    DEFAULT_PAGER               LOCALVEROCTFILEDIR
##    EXEEXT                      MAN1DIR
##    FCNFILEDIR                  MAN1EXT
##    IMAGEDIR                    MANDIR
##    INCLUDEDIR                  OCTAVE_EXEC_HOME
##    INFODIR                     OCTAVE_HOME
##    INFOFILE                    OCTDATADIR
##    LIBDIR                      OCTDOCDIR
##    LIBEXECDIR                  OCTFILEDIR
##    LOCALAPIARCHLIBDIR          OCTFONTSDIR
##    LOCALAPIFCNFILEDIR          OCTINCLUDEDIR
##    LOCALAPIOCTFILEDIR          OCTLIBDIR
##    LOCALAPIPKGDIR              STARTUPFILEDIR
##                                VERSION
## @end group
## @end example
##
## @item --link-stand-alone
## Link a stand-alone executable file.
##
## @item --mex
## Assume creation of a MEX file.  Set the default output extension to
## @file{.mex}.  Link to liboctmex instead of liboctinterp and liboctave.
##
## @item  -s
## @itemx --strip
## Strip the output file.
##
## @item  -v
## @itemx --verbose
## Echo commands as they are executed.
##
## @item file
## The file to compile or link.  Recognized file types are:
##
## @example
## @group
##    .c    C source
##    .cc   C++ source
##    .cp   C++ source
##    .cpp  C++ source
##    .CPP  C++ source
##    .cxx  C++ source
##    .c++  C++ source
##    .C    C++ source
##    .f    Fortran source (fixed form)
##    .F    Fortran source (fixed form)
##    .f90  Fortran source (free form)
##    .F90  Fortran source (free form)
##    .o    object file
##    .a    library file
## @end group
## @end example
##
## @end table
## @end deftypefn

function [output, status] = mkoctfile (varargin)

  bindir = __octave_config_info__ ("bindir");
  ext = __octave_config_info__ ("EXEEXT");

  shell_script = fullfile (bindir,
                           sprintf ("mkoctfile-%s%s", OCTAVE_VERSION, ext));

  if (! exist (shell_script, "file"))
    __gripe_missing_component__ ("mkoctfile", "mkoctfile");
  endif

  cmd = ['"' shell_script '"'];
  for i = 1:nargin
    cmd = [cmd ' "' varargin{i} '"'];
  endfor

  [sts, out] = system (cmd);

  if (nargout > 0)
    [output, status] = deal (strtrim (out), sts);
    if (sts != 0)
      warning ("mkoctfile: building exited with failure status\n");
    endif
  else
    printf ("%s", out);
    if (sts != 0)
      error ("mkoctfile: building exited with failure status\n");
    endif
  endif

endfunction
